infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java

```diff
@@ -128,32 +146,77 @@ public boolean isLRUpgradeInProgress(TxnContext txnContext) {
+                    .stream()
+                    .map(TableName::getTableName)
+                    .filter(fullName -> fullName.startsWith(V1_METADATA_TABLE_PREFIX))
+                    .map(fullName -> fullName.substring(V1_METADATA_TABLE_PREFIX.length()))
+                    .filter(remoteId -> !sinkClusterIds.contains(remoteId))
+                    .collect(Collectors.toList());
+
+            // Construct the sessions using the source and sink cluster IDs
+            for (String sourceClusterId : sourceClusterIds) {
+                sessions.addAll(sinkClusterIds.stream()
+                        .map(remoteId -> LogReplicationSession.newBuilder()
+                                .setSourceClusterId(sourceClusterId)
+                                .setSinkClusterId(remoteId)
+                                .setSubscriber(getDefaultSubscriber())
+                                .build())
+                        .collect(Collectors.toList()));
+            }
+        }
 
-        Table<LogReplicationMetadata.ReplicationEventInfoKey, LogReplicationMetadata.ReplicationEvent, Message> replicationEventTable =
-                txnContext.getTable(REPLICATION_EVENT_TABLE_NAME);
+        return sessions;
+    }
 
-        log.info("Forced snapshot sync will be triggered due to completion of rolling upgrade");
-        txnContext.putRecord(replicationEventTable, key, event, null);
+    /**
+     * Add flag to event table to trigger snapshot sync.
+     */
+    public void addSnapshotSyncEventOnUpgradeCompletion(CorfuStore corfuStore, TxnContext txnContext) {
+        for (LogReplicationSession session : buildSessionsFromOldMetadata(corfuStore, txnContext)) {
+            UUID rollingUpgradeForceSyncId = UUID.randomUUID();
+
+            // Write a rolling upgrade force snapshot sync event to the logReplicationEventTable
+            LogReplicationMetadata.ReplicationEventInfoKey key = LogReplicationMetadata.ReplicationEventInfoKey.newBuilder()
+                    .setSession(session)
+                    .build();
+
+            LogReplicationMetadata.ReplicationEvent event = LogReplicationMetadata.ReplicationEvent.newBuilder()
+                    .setEventId(rollingUpgradeForceSyncId.toString())
+                    .setType(LogReplicationMetadata.ReplicationEvent.ReplicationEventType.UPGRADE_COMPLETION_FORCE_SNAPSHOT_SYNC)
+                    .setEventTimestamp(Timestamp.newBuilder().setSeconds(Instant.now().getEpochSecond()).build())
+                    .build();
+
+            Table<LogReplicationMetadata.ReplicationEventInfoKey, LogReplicationMetadata.ReplicationEvent, Message> replicationEventTable =
+                    txnContext.getTable(REPLICATION_EVENT_TABLE_NAME);
+
+            log.info("Forced snapshot sync will be triggered due to completion of rolling upgrade for event with id: {}", rollingUpgradeForceSyncId);
+            txnContext.putRecord(replicationEventTable, key, event, null);
```

Comment:
Since the transaction context is being passed down to the event handler some of the logic would be different since we can't open a table within a transaction as well as the event types also differ, but we can move the logic to SnapshotSyncUtils to keep it all in one place.